File: MW_Test/MwSensorSummaryViewController.m
//
//  MwSensorSummaryViewController.m
//  MW_Test
//
//  Created by Rio Shao on 19/6/23.
//

#import "MwSensorSummaryViewController.h"

@interface MwSensorSummaryViewController ()

@end

@implementation MwSensorSummaryViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // initialise and get MwSensors
    _mwDevices = [[MWSensor alloc] init];
    _mwDevices = [_mwDevices getMwSensonrs];
    
    NSDate *currentDate = [NSDate date];
    NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
    [dateFormatter setDateFormat:@"yyyy_MM_dd_HHmm"];
    NSString *dateTimeString = [dateFormatter stringFromDate:currentDate];
    
    // write MwSensor Data into Csv Files
    NSURL *fileUrl = [_mwDevices writeAssessmentToFileWithFilePath:[NSString stringWithFormat:@"MwSensor_%@", dateTimeString]
                                                fileName:[NSString stringWithFormat:@"MwSensor_Summary_%@.csv", dateTimeString]];
    
    [_mwDevices writeDataToFileWithFilePath:[NSString stringWithFormat:@"MwSensor_%@", dateTimeString]
                                   fileName:[NSString stringWithFormat:@"MwSensor_Right_%@.csv", dateTimeString]
                                       type:SensorTypeRight];
    
    [_mwDevices writeDataToFileWithFilePath:[NSString stringWithFormat:@"MwSensor_%@", dateTimeString]
                                   fileName:[NSString stringWithFormat:@"MwSensor_Left_%@.csv", dateTimeString]
                                       type:SensorTypeLeft];
    
    // display CSV files into WebView
    NSMutableString *htmlContent = [self convertCsvToHtmlContent:fileUrl];
    [_csvSummary loadHTMLString:htmlContent baseURL:nil];
}

- (NSMutableString *)convertCsvToHtmlContent:(NSURL *)fileUrl {
    NSString *csvFilePath = [fileUrl path];
    NSString *csvContent = [NSString stringWithContentsOfFile:csvFilePath encoding:NSUTF8StringEncoding error:nil];

    NSMutableString *htmlContent = [NSMutableString string];
    [htmlContent appendString:@"<html><head><style>table { width: 100%; } td { padding: 5px;font-size: 32px; }</style></head><body><table border='1'>"];

    NSArray *csvRows = [csvContent componentsSeparatedByString:@"\n"];
    for (NSString *row in csvRows) {
        NSArray *rowItems = [row componentsSeparatedByString:@","];
        [htmlContent appendString:@"<tr>"];
        for (NSString *item in rowItems) {
            [htmlContent appendFormat:@"<td>%@</td>", item];
        }
        [htmlContent appendString:@"</tr>"];
    }

    [htmlContent appendString:@"</table></body></html>"];

    return htmlContent;
}

@end

File: MW_Test/MwSensorSummaryViewController.swift
import UIKit
import WebKit

@objc
class MwSensorSummaryViewController: UIViewController {
    
    let mwDevices = AssessmentSettings.sharedManager.mwSensors
    @IBOutlet var csvSummary: WKWebView!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        let fileUrl = mwDevices.writeAssessmentToFile(
            filePath: "MwSensor_\(AssessmentSettings.sharedManager.dateTimeString)",
            fileName: "MwSensor_Summary_\(AssessmentSettings.sharedManager.dateTimeString).csv"
        )
        
        // Display CSV files in WebView
        let htmlContent = convertCsvToHtmlContent(fileUrl)
        csvSummary.loadHTMLString(htmlContent, baseURL: nil)
        
    }
    
    @IBAction func discardResultsButtonPressed(_ sender: Any) {
        do {
            let folderPath = AssessmentSettings.sharedManager.filesManager.documentsDirectory.path
            + String(format: "/MwSensor_%@/",        AssessmentSettings.sharedManager.dateTimeString)
            try AssessmentSettings.sharedManager.filesManager.removeItem(atPath: folderPath)
            print("Folder and its contents deleted successfully.")
        } catch {
            print("Error deleting folder: \(error)")
        }
    }
    
    @IBAction func saveResultsPressed(_ sender: Any) {
        // Write MwSensor Data into Csv Files
        
        mwDevices.writeDataToFile(
            filePath: "MwSensor_\(AssessmentSettings.sharedManager.dateTimeString)",
            fileName: "MwSensor_Right_\(AssessmentSettings.sharedManager.dateTimeString).csv",
            type: .right
        )
        
        mwDevices.writeDataToFile(
            filePath: "MwSensor_\(AssessmentSettings.sharedManager.dateTimeString)",
            fileName: "MwSensor_Left_\(AssessmentSettings.sharedManager.dateTimeString).csv",
            type: .left
        )
    }
    
    
    func convertCsvToHtmlContent(_ fileUrl: URL) -> String {
        let csvFilePath = fileUrl.path
        let csvContent = try? String(contentsOfFile: csvFilePath, encoding: .utf8)
        
        var htmlContent = "<html><head><style>table { width: 100%; } td { padding: 5px; font-size: 32px; }</style></head><body><table border='1'>"
        
        if let csvContent = csvContent {
            let csvRows = csvContent.components(separatedBy: "\n")
            for row in csvRows {
                let rowItems = row.components(separatedBy: ",")
                htmlContent.append("<tr>")
                for item in rowItems {
                    htmlContent.append("<td>\(item)</td>")
                }
                htmlContent.append("</tr>")
            }
        }
        
        htmlContent.append("</table></body></html>")
        
        return htmlContent
    }
}
